install/kubernetes/cilium/templates/_helpers.tpl

```diff
@@ -43,7 +55,7 @@ where:
 {{- if $priorityClass }}
   {{- $priorityClass }}
 {{- else if and $root.Values.enableCriticalPriorityClass $criticalPriorityClass -}}
-  {{- if and (eq $root.Release.Namespace "kube-system") (semverCompare ">=1.10-0" $root.Capabilities.KubeVersion.Version) -}}
+  {{- if and (eq "cilium.namespace" "kube-system") (semverCompare ">=1.10-0" $root.Capabilities.KubeVersion.Version) -}}
```

Comment:
This might be problematic, checking it out

Reply:
Somehow even declaring a variable `$namespace` is not helping with `eq` comparison
```
{{- define "cilium.priorityClass" -}}
{{- $root := index . 0 -}}
{{- $priorityClass := index . 1 -}}
{{- $criticalPriorityClass := index . 2 -}}
{{- $namespace := include "cilium.namespace" . }}
{{- if $priorityClass }}
  {{- $priorityClass }}
{{- else if and $root.Values.enableCriticalPriorityClass $criticalPriorityClass -}}
  {{- if and (eq  $namespace "kube-system") (semverCompare ">=1.10-0" $root.Capabilities.KubeVersion.Version) -}}
    {{- $criticalPriorityClass }}
  {{- else if semverCompare ">=1.17-0" $root.Capabilities.KubeVersion.Version -}}
    {{- $criticalPriorityClass }}
  {{- end -}}
{{- end -}}
{{- end -}}
```

`helm template .` fails with 
```
Error: template: cilium/templates/cilium-operator/deployment.yaml:215:28: executing "cilium/templates/cilium-operator/deployment.yaml" at <include "cilium.priorityClass" (list $ .Values.operator.priorityClassName "system-cluster-critical")>: error calling include: template: cilium/templates/_helpers.tpl:54:18: executing "cilium.priorityClass" at <include "cilium.namespace" .>: error calling include: template: cilium/templates/_helpers.tpl:5:14: executing "cilium.namespace" at <.Values.namespaceOverride>: can't evaluate field Values in type []interface {}
```

Reply:
Hi @pr0PM,

Something like this should do the trick:
```go
{{- define "cilium.priorityClass" -}}
{{- $root := index . 0 -}}
{{- $priorityClass := index . 1 -}}
{{- $criticalPriorityClass := index . 2 -}}
{{- $ciliumNamespace := include "cilium.namespace" $root -}}
{{- if $priorityClass }}
  {{- $priorityClass }}
{{- else if and $root.Values.enableCriticalPriorityClass $criticalPriorityClass -}}
  {{- if and (eq $ciliumNamespace "kube-system") (semverCompare ">=1.10-0" $root.Capabilities.KubeVersion.Version) -}}
    {{- $criticalPriorityClass }}
  {{- else if semverCompare ">=1.17-0" $root.Capabilities.KubeVersion.Version -}}
    {{- $criticalPriorityClass }}
  {{- end -}}
{{- end -}}
{{- end -}}
```

I think the problem is that `.` isn't within the root context anymore because of a context change via `{{- $priorityClass := index . 1 -}}` and `{{- $criticalPriorityClass := index . 2 -}}`. That's the reason why `include "cilium.namespace" .` needs to become `include "cilium.namespace" $root` (luckily, the chart already comes with an anchor to the original root).

More details on context -- section `1.2. Passing the current and root context`: https://blog.palark.com/advanced-helm-templating/

Regards,
Philip

Reply:
Hi Philip, thanks for helping out here.
I have tested it out and it's working as expected.
The shared blog post is a gold mine, I searched a lot but wasn't able to find many resources for such issues.
Cheers!